pallets/gear-builtin/src/lib.rs

```diff
@@ -0,0 +1,368 @@
+
+        /// The builtin actor pallet id, used for deriving unique actors ids.
+        #[pallet::constant]
+        type PalletId: Get<PalletId>;
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+    impl<T: Config> Pallet<T> {
+        /// Generate an `actor_id` given a builtin ID.
+        ///
+        ///
+        /// This does computations, therefore we should seek to cache the value at the time of
+        /// a builtin actor registration.
+        pub fn generate_actor_id(builtin_id: BuiltinId) -> ProgramId {
+            hash((SEED, builtin_id).encode().as_slice()).into()
+        }
+    }
+}
+
+impl<T: Config> BuiltinRouterProvider<StoredDispatch, JournalNote, u64> for Pallet<T> {
+    type Router = BuiltinRegistry<T>;
+
+    fn provide() -> BuiltinRegistry<T> {
+        BuiltinRegistry::<T>::new()
+    }
+
+    fn provision_cost() -> u64 {
+        <T as Config>::WeightInfo::provide().ref_time()
+    }
+}
+
+pub struct BuiltinRegistry<T: Config> {
+    pub registry: BTreeMap<ProgramId, BuiltinId>,
+    pub _phantom: sp_std::marker::PhantomData<T>,
+}
+impl<T: Config> BuiltinRegistry<T> {
+    fn new() -> Self {
+        let mut registry = BTreeMap::new();
+        let mut builtin_ids = Vec::with_capacity(16);
+        <T as Config>::BuiltinActor::get_ids(&mut builtin_ids);
+        let builtin_ids_len = builtin_ids.len();
+        for builtin_id in builtin_ids {
+            let actor_id = Pallet::<T>::generate_actor_id(builtin_id);
+            registry.entry(actor_id).or_insert(builtin_id);
+        }
+        assert!(
+            registry.len() == builtin_ids_len,
+            "Duplicate builtin ids detected!"
+        );
+
+        Self {
+            registry,
+            _phantom: Default::default(),
+        }
+    }
+}
+
+impl<T: Config> BuiltinRouter for BuiltinRegistry<T> {
+    type QueuedDispatch = StoredDispatch;
+    type Output = JournalNote;
+
+    fn lookup(&self, id: &ProgramId) -> Option<BuiltinId> {
+        self.registry.get(id).copied()
+    }
+
+    fn dispatch(&self, dispatch: StoredDispatch, gas_limit: u64) -> Option<Vec<JournalNote>> {
+        let actor_id = dispatch.destination();
+
+        // Try to convert the incoming dispatch into a format that the builtin actor can handle.
+        let builtin_message =
+            <T::Message as TryFrom<StoredDispatch>>::try_from(dispatch.clone()).ok()?;
+
+        // Builtin actors can only execute dispatches of `Handle` kind and only `Handle`
+        // dispatches can end up here (TODO: elaborate).
+        if dispatch.kind() != DispatchKind::Handle {
+            unreachable!("Only handle dispatches can end up here");
+        }
+
+        let mut gas_counter = GasCounter::new(gas_limit);
+
+        let (res, gas_spent) = <T as Config>::BuiltinActor::handle(&builtin_message, gas_limit);
+
+        // We rely on a builtin actor having performed the check for gas limit consistency
```

Comment:
> for example, actor shouldn't charge almost anything if it has bench for some call A and call A is received with insufficient gas

This is exactly what is happening. The problem is that a builtin actor can execute more than one action (like with staking, for example, where you have bonding, unbonding, nominating etc.). Which one is going to be executed each time only becomes known after a message has been decoded. Hence the gas limit sufficiency check has been moved inside the actor itself. An alternative could be what it used to be in the very first implementation - we check the `gas_limit` against the max possible cost, and then refund what has been paid in excess. Or break message processing in 2 stages: decoding message and checking gas limit first, and, if success, proceed to execution.
But it would really be very twisted. The current implementation is optimal given all the constraints (provided the actor is smart enough to change the gas limit immediately after decoding a message and before launching any further processing).